### PR TITLE
feat: Add a multi-region constant B-field provider

### PR DESCRIPTION
**Pull request by stephenswat**
In order to facilitate the ALICE3 use-case (as described to me by @Cas1997) we need to support magnetic fields that are constant in multiple regions, but which may have different field values in each of these regions. I implemented a new magnetic field provider based on a mapping of 3D ranges to field vectors which should support this use case.

The core idea is that the user provides a list of regions and the provider will check (with some simple caching) which region matches. The mechanism implemented is that -- in case of overlap -- the region later in the list takes precedence.

Also implements C++ and Python tests for this new field provider.

## Release Notes

- **New Features**
	- Introduced the `MultiRangeBField` class for modeling magnetic fields with multiple overlapping regions.
	- Added the `RangeXDDim3` class for handling three-dimensional ranges in the geometry module.

- **Tests**
	- Implemented comprehensive tests for the `MultiRangeBField` class, covering various scenarios to validate its behavior.
	- Added a new test function for the `acts.MultiRangeBField` class in Python.

- **Documentation**
	- Updated documentation to include details about the `Acts::MultiRangeBField` class and its functionality, including caching mechanisms.

- **Chores**
	- Expanded the unit test suite to include tests for the `MultiRangeBField` class.
	- Added a new source file for the `MultiRangeBField` class to the project.

### CODE REVIEW

File: Core/include/Acts/MagneticField/MultiRangeBField.hpp [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+ public:
+  /// @{
+  /// @brief Construct a magnetic field from a vector of ranges.
+  ///
+  /// @warning These ranges are listed in increasing order of precedence,
+  /// i.e. ranges further along the vector have higher priority.
+  MultiRangeBField(const std::vector<BFieldRange>& ranges) : m_ranges(ranges) {}
+
+  MultiRangeBField(std::vector<BFieldRange>&& ranges) : m_ranges(ranges) {}
```
[the remark this answers is not in the document]
Comment (stephenswat) [reply Core/include/Acts/MagneticField/MultiRangeBField.hpp:45]: Shit the robots really are going to take our jobs.

**Comment by stephenswat** [comment]
> It would be nice to support this in covfie field as well or does it already?

Of course, that's a good idea. :+1:

File: Core/include/Acts/MagneticField/MultiRangeBField.hpp [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+#include "Acts/MagneticField/MagneticFieldProvider.hpp"
+#include "Acts/Utilities/RangeXD.hpp"
+
+namespace Acts {
+
+/// @ingroup MagneticField
+///
+/// @brief Magnetic field provider modelling a magnetic field consisting of
+/// several (potentially overlapping) regions of constant values.
+class MultiRangeBField final : public MagneticFieldProvider {
```
Comment (paulgessinger) [inline Core/include/Acts/MagneticField/MultiRangeBField.hpp:22]: Can you split this up into a cpp file?

**Review by paulgessinger — COMMENTED** [review]
We could even consider making the regular `ConstantBField` use this behind the scenes, since it's essentially a special case.

File: Core/include/Acts/MagneticField/MultiRangeBField.hpp (hunk 2 above)
Comment (stephenswat) [reply Core/include/Acts/MagneticField/MultiRangeBField.hpp:22]: Done, thanks.

File: Core/src/MagneticField/MultiRangeBField.cpp [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  return MagneticFieldProvider::Cache(std::in_place_type<Cache>, mctx);
+}
+
+Result<Vector3> MultiRangeBField::getField(
+    const Vector3& position, MagneticFieldProvider::Cache& cache) const {
+  // Because we assume that the regions are in increasing order of
+  // precedence, we can iterate over the array, remembering the _last_
+  // region that contained the given point. At the end of the loop, this
+  // region will then be the one we query for its field vector.
+  std::optional<std::size_t> found_range = {};
```
Comment (paulgessinger) [inline Core/src/MagneticField/MultiRangeBField.cpp:33]: Nitpick: we usually use `camelCase`.

**Review by paulgessinger — COMMENTED** [review]
I have one more nitpick.

**Review by paulgessinger — APPROVED** [review]

**Comment by andiwand** [comment]
resolved the AI conversations the make this pass

**Comment by acts-project-service** [comment]
## :red_circle: Athena integration test results [b3933b06cd9722f52f568cae3513123f343923e4]
### Build job with this PR failed!
Please investigate the [build job](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/47603352) for the [pipeline](https://gitlab.cern.ch/acts/acts-athena-ci/-/pipelines/9334078)!

**Pull request merged**